raiden-ts/src/channels/epics.ts

```diff
@@ -798,71 +815,72 @@ export const channelDepositEpic = (
   }: RaidenEpicDeps,
 ): Observable<channelDeposit.failure> =>
   action$.pipe(
-    filter(isActionOf([channelDeposit.request, channelOpen.request])),
-    // ignore channelOpen.request if no deposit is requested on it
-    filter((action) => !channelOpen.request.is(action) || !!action.payload.deposit?.gt?.(0)),
-    groupBy((action) => action.meta.tokenNetwork), // serialize per tokenNetwork
+    filter(isActionOf(channelDeposit.request)),
+    groupBy((action) => action.meta.tokenNetwork),
     mergeMap((grouped$) =>
       grouped$.pipe(
-        mergeMap((action) => {
-          const error = new RaidenError(ErrorCodes.CNL_NO_OPEN_CHANNEL_FOUND, action.meta);
-          const notOpenAction = channelDeposit.failure(error, action.meta);
-          // this 'merge' re-emits the requests as soon as the channel is detected as 'open';
-          // emits deposit failure if channel isn't open or open request fails
-          return merge(
-            latest$.pipe(
-              pluck('state', 'channels', channelKey(action.meta)),
-              // if it's a open request, wait confirmed channel, otherwise pass to validate now
-              filter((channel) => (channelOpen.request.is(action) ? isntNil(channel) : true)),
-              map((channel) => {
-                if (channel?.state !== ChannelState.open) return notOpenAction;
-                return action; // re-emit deposit/open action when channel is open
-              }),
-            ),
-            action$.pipe(
-              filter(channelOpen.failure.is),
-              filter(
-                (failure) =>
-                  failure.meta.tokenNetwork === action.meta.tokenNetwork &&
-                  failure.meta.partner === action.meta.partner,
-              ),
-              mapTo(notOpenAction),
-            ),
-          ).pipe(take(1)); // for each request, take only first re-emited request or failure
-        }),
-        // concatMap to deposit on channels serially per tokenNetwork, in the order they're
-        // detected as 'open' and confirmed
+        // groupBy + concatMap ensure actions handling is serialized in a given tokenNetwork
```

Comment:
Shouldn't we group by channel only? Could be that we open a channel with an initial deposit. While we wait for the on-chain transactions to open before we can deposit, we could not deposit to an already open different channel in parallel. At least how I understand it.

Reply:
No, that wouldn't work, because both requests would fight for the `allowance` resource: `approve` needs to be called before `setTotalDeposit`, and `setTotalDeposit` also decreases allowance, and as they go on-chain, races could make us waste gas on unneeded transactions and require us to `retryTx` something which should work, so it needs to be serialized per token. Fortunatelly, this only happens for parallel/concurrent deposits on the same tokenNetwork, which usually doesn't happen outside of scenarios, and for them, it's ok to serialize and make it wait

Reply:
Hmm okay.